File: app/src/main/java/com/android/sunshine/app/fragments/DetailFragment.java
package com.android.sunshine.app.fragments;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.CardView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.android.sunshine.app.App;
import com.android.sunshine.app.R;
import com.android.sunshine.app.activities.DetailActivity;
import com.android.sunshine.app.utils.Navigator;
import com.android.sunshine.app.utils.WeatherDetails;
import java.util.Locale;
import javax.inject.Inject;

public class DetailFragment extends Fragment {

    public static final String DETAIL_TRANSITION_ANIMATION = "DTA";
    public static final String DETAIL_URI = "URI";

    @Bind(R.id.detail_icon) ImageView iconView;
    @Bind(R.id.detail_date_textview) TextView dateView;
    @Bind(R.id.detail_forecast_textview) TextView mDescriptionView;
    @Bind(R.id.detail_high_textview) TextView mHighTempView;
    @Bind(R.id.detail_low_textview) TextView mLowTempView;
    @Bind(R.id.detail_humidity_textview) TextView mHumidityView;
    @Bind(R.id.detail_wind_textview) TextView mWindView;
    @Bind(R.id.detail_pressure_textview) TextView mPressureView;

    @Inject
    Navigator navigator;

    private boolean transitionAnimation;
    private String weatherData;

    public DetailFragment() {
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        ((App) getActivity().getApplication()).getComponent().inject(this);
        Bundle arguments = getArguments();
        if (arguments != null) {
            transitionAnimation = arguments.getBoolean(DetailFragment.DETAIL_TRANSITION_ANIMATION, false);
        }

        final View view = inflater.inflate(R.layout.fragment_detail_start, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        final WeatherDetails weatherDetails = getArguments().getParcelable(Navigator.WEATHER_DETAILS);
        renderData(weatherDetails);
    }

    @Override
    public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
        if (getActivity() instanceof DetailActivity) {
            inflater.inflate(R.menu.detail_fragment, menu);
            finishCreatingMenu(menu);
        }
    }

    private void renderData(final WeatherDetails weather) {
        ViewParent vp = getView().getParent();
        if (vp instanceof CardView) {
            ((View) vp).setVisibility(View.VISIBLE);
        }

        dateView.setText(weather.getDate());
        mDescriptionView.setText(weather.getDescription());
        mHighTempView.setText(weather.getMax());
        mLowTempView.setText(weather.getMin());
        mHumidityView.setText(weather.getHumidity());
        mWindView.setText(weather.getWind());
        mPressureView.setText(weather.getPressure());
        iconView.setImageResource(weather.getIconResourceId());

        weatherData = String.format(Locale.getDefault(), "%s - %s - %s/%s", weather.getDate(), weather.getDescription(), weather.getMax(), weather.getMin());

        AppCompatActivity activity = (AppCompatActivity) getActivity();
        Toolbar toolbarView = (Toolbar) getView().findViewById(R.id.toolbar);

        // We need to start the enter transition after the data has loaded
        if (transitionAnimation) {
            activity.supportStartPostponedEnterTransition();

            if (null != toolbarView) {
                activity.setSupportActionBar(toolbarView);

                activity.getSupportActionBar().setDisplayShowTitleEnabled(false);
                activity.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            }
        } else {
            if (null != toolbarView) {
                Menu menu = toolbarView.getMenu();
                if (null != menu) {
                    menu.clear();
                }
                toolbarView.inflateMenu(R.menu.detail_fragment);
                finishCreatingMenu(toolbarView.getMenu());
            }
        }
    }

    private void finishCreatingMenu(final Menu menu) {
        MenuItem menuItem = menu.findItem(R.id.action_share);
        menuItem.setIntent(navigator.createShareIntent(weatherData));
    }
}

File: app/src/main/kotlin/com/android/sunshine/app/fragments/DetailFragment.kt
package com.android.sunshine.app.fragments

import android.app.Fragment
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.LoaderManager
import android.support.v4.content.CursorLoader
import android.support.v4.content.Loader
import android.support.v4.view.MenuItemCompat
import android.support.v7.widget.ShareActionProvider
import android.view.*
import android.widget.ImageView
import android.widget.TextView
import com.android.sunshine.app.R
import com.android.sunshine.app.activities.DetailActivity
import com.android.sunshine.app.model.WeatherContract
import com.android.sunshine.app.utils.Utilities
import kotlinx.android.synthetic.fragment_detail.*

import java.util.Locale

public class DetailFragment : Fragment(), LoaderManager.LoaderCallbacks<Cursor> {

    var location: String = ""
    var shareIntent: Intent? = null
    var weatherData: String = ""

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super<Fragment>.onActivityCreated(savedInstanceState)
        setHasOptionsMenu(true)

        if (savedInstanceState != null) {
            location = savedInstanceState.getString(LOCATION_KEY, "")
        }
        val arguments = getArguments()
        if (arguments != null && arguments.containsKey(DetailActivity().DATE_KEY)) {
            getLoaderManager().initLoader<Cursor>(DETAIL_LOADER, null, this)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view = inflater.inflate(R.layout.fragment_detail, container, false)
        return view
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super<Fragment>.onSaveInstanceState(outState)
        if (location.isNotEmpty()) {
            outState!!.putString(LOCATION_KEY, location)
        }
    }

    override fun onResume() {
        super<Fragment>.onResume()
        val arguments = getArguments()
        if (arguments != null && !location.equals(Utilities.getLocationSettings(getActivity())) && arguments.containsKey(DetailActivity().DATE_KEY)) {
            getLoaderManager().restartLoader<Cursor>(DETAIL_LOADER, null, this)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.detail_fragment, menu)
        val menuItem = menu!!.findItem(R.id.menu_item_share)
        val actionProvider = MenuItemCompat.getActionProvider(menuItem) as ShareActionProvider
        if (shareIntent == null) {
            createShareIntent()
        }
        actionProvider.setShareIntent(shareIntent)
    }

    override fun onCreateLoader(id: Int, args: Bundle): Loader<Cursor> {
        val date = getArguments().getString(DetailActivity().DATE_KEY)
        location = Utilities.getLocationSettings(getActivity())
        val weatherUri = WeatherContract.WeatherEntry.buildWeatherLocationWithDate(location, date)
        return CursorLoader(getActivity(), weatherUri, COLUMNS, null, null, WeatherContract.WeatherEntry.COLUMN_DATETEXT + " ASC")
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        if (data.moveToFirst()) {
            val isMetric = Utilities.isMetric(getActivity())
            val weatherId = data.getInt(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_WEATHER_ID))
            val description = data.getString(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_SHORT_DESC))
            val date = data.getString(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_DATETEXT))
            val wind = data.getString(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_WIND_SPEED))
            val pressure = data.getString(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_PRESSURE))
            val humidity = data.getString(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_HUMIDITY))
            val maxTemp = data.getDouble(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_MAX_TEMP))
            val minTemp = data.getDouble(data.getColumnIndex(WeatherContract.WeatherEntry.COLUMN_MIN_TEMP))
            val max = Utilities.formatTemperature(getActivity(), maxTemp, isMetric).toString()
            val min = Utilities.formatTemperature(getActivity(), minTemp, isMetric).toString()

            detail_date.setText(date)
            detail_forecast!!.setText(description)
            detail_max!!.setText(max)
            detail_min!!.setText(min)
            detail_humidity!!.setText(humidity)
            detail_wind!!.setText(wind)
            detail_pressure!!.setText(pressure)
            detail_day.setText(Utilities.getDayName(getActivity(), date))
            detail_icon!!.setImageResource(Utilities.getArtResourceForWeatherCondition(weatherId))

            weatherData = String().format(Locale.getDefault(), "%s - %s - %s/%s", date, description, max, min)
        }
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {
    }

    private fun createShareIntent() {
        shareIntent = Intent(Intent.ACTION_SEND)
        shareIntent!!.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET)
        shareIntent!!.setType("text/plain")
        shareIntent!!.putExtra(Intent.EXTRA_TEXT, weatherData + " #sunshine")
    }

    companion object {

        public val DETAIL_LOADER: Int = 0
        public val LOCATION_KEY: String = "location"
        private val COLUMNS = array(WeatherContract.WeatherEntry.TABLE_NAME + "._id", WeatherContract.WeatherEntry.COLUMN_DATETEXT, WeatherContract.WeatherEntry.COLUMN_SHORT_DESC,
                WeatherContract.WeatherEntry.COLUMN_MAX_TEMP, WeatherContract.WeatherEntry.COLUMN_MIN_TEMP, WeatherContract.WeatherEntry.COLUMN_HUMIDITY, WeatherContract.WeatherEntry.COLUMN_PRESSURE,
                WeatherContract.WeatherEntry.COLUMN_WIND_SPEED, WeatherContract.WeatherEntry.COLUMN_WEATHER_ID)
    }
}